compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -125,11 +125,15 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                     self.assemble_async_iterator_candidates(obligation, &mut candidates);
                 } else if tcx.is_lang_item(def_id, LangItem::AsyncFnKindHelper) {
                     self.assemble_async_fn_kind_helper_candidates(obligation, &mut candidates);
+                } else if tcx.is_lang_item(def_id, LangItem::AsyncFn)
+                    || tcx.is_lang_item(def_id, LangItem::AsyncFnOnce)
+                    || tcx.is_lang_item(def_id, LangItem::AsyncFnMut)
+                {
+                    self.assemble_async_closure_candidates(obligation, &mut candidates);
```

Comment:
Moving `assemble_async_closure_candidates` here isn't necessary as part of the PR, but it is part of the FIXME below. I'm happy to leave this change or take it out in the interest of keeping a more focused PR.